fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/domain/SearchQueryRenderer.java

```diff
@@ -1341,23 +1348,49 @@ protected String getLRAlias(int aliasIndex) {
      * @param paramAlias
      * @return
      */
-    protected String getTokenParamTable(ExpNode filter, String resourceType, String paramAlias) {
-        ColumnExpNodeVisitor visitor = new ColumnExpNodeVisitor(); // gathers all columns used in the filter expression
-        Set<String> columns = filter.visit(visitor);
-        boolean usesTokenValue = columns.contains(DataDefinitionUtil.getQualifiedName(paramAlias, TOKEN_VALUE)) ||
-                                    columns.contains(DataDefinitionUtil.getQualifiedName(paramAlias, CODE_SYSTEM_ID));
+    protected String getTokenParamTable(ExpNode filter, String resourceType, String paramAlias, boolean filterIsOptimized) {
 
         final String xxTokenValues;
-        if (usesTokenValue) {
-            // can't optimize because we filter on TOKEN_VALUE
-            xxTokenValues = resourceType + "_TOKEN_VALUES_V";
+        final String xxResourceTokenRefs = resourceType + "_RESOURCE_TOKEN_REFS";
+        if (filterIsOptimized) {
+            // only filters on COMMON_TOKEN_VALUE_ID so we can optimize by using only the
+            // xx_resource_token_refs table. This results in much better cardinality
+            // estimation and an a better plan
+            xxTokenValues = xxResourceTokenRefs;
         } else {
-            // only filters on COMMON_TOKEN_VALUE_ID so we can optimize
-            xxTokenValues = resourceType + "_RESOURCE_TOKEN_REFS";
+            // can't optimize because we filter on TOKEN_VALUE. If the schema type is distributed,
+            // this becomes a special case because COMMON_TOKEN_VALUES is distributed on TOKEN_VALUE which
+            // is not a common column in the join predicate. In this case, we have to use a WITH clause
+            // which then gets joined with the parameter table (Citus will aggregate the results of the WITH
+            // select statement on the coordinator then pass those results to the worker nodes
+            if (schemaType == SchemaType.DISTRIBUTED) {
+                xxTokenValues = xxResourceTokenRefs;
+            } else {
+                // just use the standard view
+                xxTokenValues = resourceType + "_TOKEN_VALUES_V";
+            }
         }
         return xxTokenValues;
     }
 
+    /**
+     * Inspect the filter expression to see if it mentions the TOKEN_VALUE
+     * or CODE_SYSTEM_ID columns. If neither of these columns is not mentioned, 
```

Comment:
the double-negative makes me wonder if this is right

Reply:
ok, removed the not - which makes the comment now match the actual implementation (which is correct).